### ISSUE TITLE
TinyLFU

### ISSUE THREAD

**Issue by ben-manes** [jedisct1/rust-arc-cache#1]
Can you give this [policy](http://highscalability.com/blog/2016/1/25/design-of-a-modern-cache.html) a try? I'd be interested in hearing feedback, concerns, ideas, etc.

**Comment by jedisct1** [comment]
I'm implementing TinyLFU as we speak :)

**Comment by ben-manes** [comment]
If it helps see my [CountMinSketch](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java) and [BloomFilter](https://github.com/ben-manes/caffeine/blob/master/simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/membership/bloom/BloomFilter.java) (32-bit hash due to Java's `hashCode()`). There is also the [go-tinylfu](https://github.com/dgryski/go-tinylfu), which I think you saw, that is very readable. I think writing sketches is pretty fun part of this algorithm.

**Comment by jedisct1** [comment]
My current use case is a [DNS server](https://github.com/jedisct1/edgedns), where the primary focus of the cache policy is not the hit ratio. Latency and adversarial defense are more important.

Resetting the TinyLFU counters is a blocking operation, and may be a showstopper for that use case.

**Comment by ben-manes** [comment]
In my [simulations](https://docs.google.com/spreadsheets/d/1G3deNz1gJCoXBE2IuraUSwLE7H_EMn4Sn2GU0HTpI5Y/edit?usp=sharing) an incremental reset performed equally well. So while it increases the error, it doesn't really diminish whether an entry is a heavy hitter. Those entries will quickly max out the frequency, or we end up taking a warm (but not hot) candidate. Then the algorithm is `O(1)` and not merely amortized `O(1)`. I used periodic reset since that can easily be vectorized or, in my case, deferred to be async.

In the implementation, I also added HashDoS [protection](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L634) by randomly accepting a warm candidate. That disrupts an attack trying to artificially raise the victim's frequency due to a collision, thereby halting admissions.
